File: lc_soln/0239_sliding_window_maximum/solution.cs
public class Solution
{
    public int[] MaxSlidingWindow(int[] nums, int k)
    {
        var q = new LinkedList<int>();
        var result = new List<int>();
        int start = 0;
        for (int i = 0; i < nums.Length; ++i)
        {
            while (q.Count > 0 && nums[q.Last.Value] < nums[i])
            {
                q.RemoveLast();
            }
            q.AddLast(i);
            if (q.First.Value < start)
            {
                q.RemoveFirst();
            }
            if (i - start + 1 >= k)
            {
                result.Add(nums[q.First.Value]);
                start++;
            }
        }
        return result.ToArray();
    }
}

File: lc_soln/0239_sliding_window_maximum/solution.py
from collections import deque
from typing import List


class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        q = deque()
        result = []
        start = 0
        for i, num in enumerate(nums):
            while q and nums[q[-1]] < num:
                q.pop()
            q.append(i)

            if start > q[0]:
                q.popleft()

            if i - start + 1 >= k:
                result.append(nums[q[0]])
                start += 1

        return result


def main(): ...


if __name__ == "__main__":
    main()
